### PR TITLE
refactor NodeRole so extensions can participate in disco and announcement

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
Druid is like, absurdly extensible, which is pretty cool. Extensions are able to create entire new standalone tools and even full services, and they can be quite powerful because they have access to the wealth of facilities provided by the Druid codebase. However, they are not currently able to announce or discover themselves because we use a `NodeRole` enumeration which limits participation to known service types.

This PR modifies `NodeRole` to instead be a class, mimicking the contract the enum provided so that there is no apparent change, and allowing custom `NodeRoles` to be used alongside the built-in service types. I've added an integration test that tests discovery and self discovery (`/status/selfDiscovered`, which wasn't previously covered by tests afaict) of both the built-in node types as well as an additional custom node-role service extension.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/discovery/NodeRole.java [hunk 1]
```diff
@@ -31,33 +38,88 @@
  * These abstractions can probably be merged when Druid updates to Jackson 2.9 that supports JsonAliases, see
  * see https://github.com/apache/druid/issues/7152.
  */
-public enum NodeRole
+public class NodeRole
```
Comment (gianm) [inline server/src/main/java/org/apache/druid/discovery/NodeRole.java:41]: Please edit the javadoc:

- It's no longer an enum, the the javadoc says it is.
- The javadoc should explain the difference between "name" and "jsonName" and why they both exist.

Comment (clintropolis) [reply server/src/main/java/org/apache/druid/discovery/NodeRole.java:50]: thanks, forgot to update, should be good now

**Review by jihoonson — APPROVED** [review]
LGTM

**Review by maytasm — APPROVED** [review]

**Comment by clintropolis** [comment]
thanks for review @gianm, @jihoonson, and @maytasm :metal:

**Pull request merged**